File: day-04/go/part1.go
package main

func solvePart1(grid [][]string) int {
	count := 0

	for r := 0; r < len(grid); r++ {
		for c := 0; c < len(grid[r]); c++ {
			if grid[r][c] != "@" {
				continue
			}

			adjacentItems := 0
			for _, pos := range adjacentItemPositions {
				if checkItem(grid, r+pos[0], c+pos[1]) {
					adjacentItems += 1
				}
			}
			if adjacentItems < 4 {
				count += 1
			}
		}
	}
	return count
}

File: day-04/ts/part1.ts
import { checkItem } from "./util.js";

export function part1(lines: string[]) {
  const grid: string[][] = [];
  lines.forEach((line) => grid.push(line.split("")));
  let count = 0;

  for (let r = 0; r < grid.length; r++) {
    for (let c = 0; c < grid[r]!.length; c++) {
      if (grid[r]![c] !== "@") continue;

      let adjacentItems = 0;
      if (checkItem(grid, r - 1, c - 1)) adjacentItems += 1;
      if (checkItem(grid, r - 1, c)) adjacentItems += 1;
      if (checkItem(grid, r - 1, c + 1)) adjacentItems += 1;
      if (checkItem(grid, r, c - 1)) adjacentItems += 1;
      if (checkItem(grid, r, c + 1)) adjacentItems += 1;
      if (checkItem(grid, r + 1, c - 1)) adjacentItems += 1;
      if (checkItem(grid, r + 1, c)) adjacentItems += 1;
      if (checkItem(grid, r + 1, c + 1)) adjacentItems += 1;

      if (adjacentItems < 4) count += 1;
    }
  }

  return count;
}
